app/controllers/concerns/table_on_steroids/table_concern.rb

```diff
@@ -36,168 +38,185 @@ module TableConcern
+          params[:filters].each_pair do |k, v|
             filter = columns_on_steroid[k]
             next unless filter.present? && filter[t].present?
+
             objects = filter[t][:filter_lambda].call(objects, v)
           end
         end
 
-        #order
-        if params[:knowledge] && params[:knowledge] && params[:knowledge][:order].present? && (object_order = columns_on_steroid[params[:knowledge][:order]]).present?       
-          if(object_order[t] && object_order[t][:order_lambda])
-            objects = objects.reorder(nil) if(objects.is_a?(ActiveRecord::Base) || objects.is_a?(ActiveRecord::Relation))
+        # order
+        if params[:knowledge] && params[:knowledge] && params[:knowledge][:order].present? && (object_order = columns_on_steroid[params[:knowledge][:order]]).present?
+          if object_order[t] && object_order[t][:order_lambda]
+            objects = objects.reorder(nil) if objects.is_a?(ActiveRecord::Base) || objects.is_a?(ActiveRecord::Relation)
             objects = object_order[t][:order_lambda].call(objects)
           end
-        elsif  ( object_order = columns_on_steroid.select{ |c,v| v[t] && v[t][:default_order] }.first&.last).present?
-          if(object_order[t] && object_order[t][:order_lambda])  
-            objects = objects.reorder(nil) if(objects.is_a?(ActiveRecord::Base) || objects.is_a?(ActiveRecord::Relation))
+        elsif (object_order = columns_on_steroid.select { |_c, v| v[t] && v[t][:default_order] }.first&.last).present?
+          if object_order[t] && object_order[t][:order_lambda]
+            objects = objects.reorder(nil) if objects.is_a?(ActiveRecord::Base) || objects.is_a?(ActiveRecord::Relation)
             objects = object_order[t][:order_lambda].call(objects)
           end
-        elsif(objects.is_a?(ActiveRecord::Base) || objects.is_a?(ActiveRecord::Relation))
-          #objects = objects.order('created_at desc')
+        elsif objects.is_a?(ActiveRecord::Base) || objects.is_a?(ActiveRecord::Relation)
+          # objects = objects.order('created_at desc')
         end
-
       end
 
-      #pagination 
+      # pagination
       return (include_counts ? [objects, 1, objects.count] : objects) if all_pages
-      if(objects.is_a?(ActiveRecord::Base) || objects.is_a?(ActiveRecord::Relation))
-        objects = objects.page(params[:page]).per(OBJECTS_PER_PAGE) 
+
+      unpaginated_objects = objects
+      if objects.is_a?(ActiveRecord::Base) || objects.is_a?(ActiveRecord::Relation)
+        objects = objects.page(params[:page]).per(OBJECTS_PER_PAGE)
         total_pages = objects.total_pages
         total_count = objects.total_count
-      else 
+      else
         objects = Kaminari.paginate_array(objects).page(params[:page]).per(OBJECTS_PER_PAGE)
         total_pages = objects.total_pages
         total_count = objects.total_count
       end
+      # fetch_extra object if needed
+
+      if (fetch_object_id = opts[:fetch_object_id]).present? &&
```

Comment:
^ with the above you wouldn't need opts[:fetch_object_search_lambda] or opts[:key_lambda] and you would already have the correct page so you wouldn't need to check here